Hoist all bookkeeping to global level

jussisaurio opened the pull request:
right before falling asleep last night i thought about what it would take to make this library support `useEffect` calls outside a component. i thought it might work if everything inside the closure of `createState` was hoisted out of it: i.e. make all the bookkeeping integers, sets and maps global & make `deferredUpdate`, `trackedExecute`, `wrap` etc. all global as well. This also removes the need for any of the StateInternals from the proxies except STATE_FLAG.

`useEffect` outside a component just runs it every time dependencies change, and skips everything related to `componentEffects` (which I guess is just needed for running the cleanup functions when a component unmounts.

not entirely sure if this breaks something -- tests are green, but maybe requires writing some more elaborate test cases to ensure nothing goes wrong

jussisaurio committed:
Hoist all bookkeeping to global level

Turns out that by hoisting all the bookkeeping variables
and tracking functions to the module global level, all
existing functionality is retained while also allowing
behavior such as running useEffect outside components.

jussisaurio committed:
Add test for two separate render roots

oamaok commented:
Would it be reasonable to continue this approach so that the `state` argument can be dropped completely from the `render` call?

jussisaurio committed:
Remove StateStack

Passing a state object to the render call
is now not necessary, and the StateStack
bookkeeping variable can also be entirely
removed.

jussisaurio commented:
Yeah turns out we can do that. I'd like to battle test this with something a bit more substantial like [this](https://github.com/oamaok/modulate) to get a better feel for whether or not something has actually been broken here. If not, then this seems like a decent simplification overall.

jussisaurio commented:
> Yeah turns out we can do that. I'd like to battle test this with something a bit more substantial like [this](https://github.com/oamaok/modulate) to get a better feel for whether or not something has actually been broken here. If not, then this seems like a decent simplification overall.

Seems to work, although I am creating some absolutely horrific sounds
